File: package/main/src/Object/deepClone.ts
/**
 * Recursively clones a value.
 */
const cloneValue = (value: unknown, depth: number): unknown => {
  if (value === null || typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    const result: unknown[] = [];
    for (const element of value) {
      result.push(cloneValue(element, depth + 1));
    }
    return result;
  }

  if (value instanceof Date) {
    return new Date(value);
  }

  if (value instanceof RegExp) {
    return new RegExp(value.source, value.flags);
  }

  if (value instanceof Map) {
    const result = new Map();
    for (const [k, v] of value) {
      result.set(cloneValue(k, depth + 1), cloneValue(v, depth + 1));
    }
    return result;
  }

  if (value instanceof Set) {
    const result = new Set();
    for (const v of value) {
      result.add(cloneValue(v, depth + 1));
    }
    return result;
  }

  // Plain object
  const result: Record<string, unknown> = {};
  for (const key of Object.keys(value as Record<string, unknown>)) {
    result[key] = cloneValue(
      (value as Record<string, unknown>)[key],
      depth + 1,
    );
  }
  return result;
};

/**
 * Creates a deep clone of the given value.
 *
 * @param value - The value to deep clone.
 * @returns A deep clone of the input value.
 *
 * @example
 * ```typescript
 * const original = { a: 1, b: { c: 2 } };
 * const cloned = deepClone(original);
 * cloned.b.c = 99;
 * original.b.c; // still 2
 * ```
 *
 * @remarks
 * **Prototype pollution warning:** This function does not filter out
 * prototype-polluting keys (`__proto__`, `constructor`, `prototype`).
 * If processing user-controlled input, sanitize with the appropriate
 * `removePrototype*` helper before calling this function:
 * - `removePrototype` — shallow sanitization of a single object
 * - `removePrototypeDeep` — recursive sanitization of a single object (for deeply nested data)
 * - `removePrototypeMap` — shallow sanitization of an array of objects
 * - `removePrototypeMapDeep` — recursive sanitization of an array of objects (for deeply nested data)
 */
export const deepClone = <T>(value: T): T => {
  if (value === null || typeof value !== "object") {
    return value;
  }
  return cloneValue(value, 0) as T;
};

File: package/umt_python/src/object/deep_clone.py
import re
from copy import copy
from datetime import date, datetime, time, timedelta
from typing import Any


def _clone_value(value: Any) -> Any:  # noqa: ANN401
    """Recursively clones a value."""
    if value is None or isinstance(value, (int, float, str, bool, bytes)):
        return value

    if isinstance(value, list):
        return [_clone_value(element) for element in value]

    if isinstance(value, tuple):
        return tuple(_clone_value(element) for element in value)

    if isinstance(value, (datetime, date, time, timedelta)):
        return copy(value)

    if isinstance(value, re.Pattern):
        return re.compile(value.pattern, value.flags)

    if isinstance(value, dict):
        result: dict[str, Any] = {}
        for key in value:
            result[key] = _clone_value(value[key])
        return result

    if isinstance(value, set):
        return {_clone_value(v) for v in value}

    if isinstance(value, frozenset):
        return frozenset(_clone_value(v) for v in value)

    return copy(value)


def deep_clone(value: Any) -> Any:  # noqa: ANN401
    """
    Creates a deep clone of the given value.

    Recursively copies nested structures including dicts, lists, tuples,
    sets, frozensets, datetime objects, and compiled regex patterns.
    Primitive values (int, float, str, bool, None, bytes) are returned as-is.

    Args:
        value: The value to deep clone.

    Returns:
        A deep clone of the input value.

    Example:
        >>> original = {"a": 1, "b": {"c": 2}}
        >>> cloned = deep_clone(original)
        >>> cloned["b"]["c"] = 99
        >>> original["b"]["c"]
        2
    """
    if value is None or isinstance(value, (int, float, str, bool, bytes)):
        return value
    return _clone_value(value)
